internal/cloud/backend.go

```diff
@@ -1160,39 +1170,75 @@ func (b *Cloud) cliColorize() *colorstring.Colorize {
 	}
 }
 
-func (b *Cloud) workspaceTagsRequireUpdate(ctx context.Context, workspace *tfe.Workspace, workspaceMapping WorkspaceMapping) (bool, error) {
-	if workspaceMapping.Strategy() == WorkspaceTagsStrategy {
-		existingTags := map[string]struct{}{}
-		for _, t := range workspace.TagNames {
-			existingTags[t] = struct{}{}
-		}
+type tagRequiresUpdateResult struct {
+	requiresUpdate bool
+	supportsKVTags bool
+}
 
-		for _, tag := range workspaceMapping.TagsAsSet {
-			if _, ok := existingTags[tag]; !ok {
-				return true, nil
-			}
-		}
-	} else if workspaceMapping.Strategy() == WorkspaceKVTagsStrategy {
-		existingTags := make(map[string]string)
-		bindings, err := b.client.Workspaces.ListTagBindings(ctx, workspace.ID)
-		if err != nil && err == tfe.ErrResourceNotFound {
-			return true, ErrCloudDoesNotSupportKVTags
-		} else if err != nil {
-			return true, err
-		}
+func (b *Cloud) workspaceTagsRequireUpdate(ctx context.Context, workspace *tfe.Workspace, workspaceMapping WorkspaceMapping) (result tagRequiresUpdateResult, err error) {
+	result = tagRequiresUpdateResult{
+		supportsKVTags: true,
+	}
 
-		for _, binding := range bindings {
-			existingTags[binding.Key] = binding.Value
+	// First, depending on the strategy, build a map of the tags defined in config
+	// so we can compare them to the actual tags on the workspace
+	normalizedTagMap := make(map[string]string)
+	if workspaceMapping.IsTagsStrategy() {
+		for _, b := range workspaceMapping.asTFETagBindings() {
+			normalizedTagMap[b.Key] = b.Value
 		}
-
-		for tag, val := range workspaceMapping.TagsAsMap {
-			if existingVal, ok := existingTags[tag]; !ok || existingVal != val {
-				return true, nil
+	} else {
+		// Not a tag strategy
+		return
+	}
+
+	// Fetch tag bindings and determine if they should be checked
+	bindings, err := b.client.Workspaces.ListTagBindings(ctx, workspace.ID)
+	if err != nil && errors.Is(err, tfe.ErrResourceNotFound) {
+		// By this time, the workspace should have been fetched, proving that the
+		// authenticated user has access to it. If the tag bindings are not found,
+		// it would mean that the backened does not support tag bindings.
+		result.supportsKVTags = false
+	} else if err != nil {
+		return
+	}
+
+	err = nil
+check:
+	// Check desired workspace tags against existing tags
+	for k, v := range normalizedTagMap {
+		log.Printf("[TRACE] cloud: Checking tag %q=%q", k, v)
+		if v == "" {
+			// Tag can exist in legacy tags or tag bindings
+			if !slices.Contains(workspace.TagNames, k) || (result.supportsKVTags && !slices.ContainsFunc(bindings, func(b *tfe.TagBinding) bool {
```

Comment:
Consider extracting and assigning the slice functions locally or elsewhere, for readability?